test-suite/src/test/scala/org/scalajs/testsuite/compiler/OptimizerTest.scala

```diff
@@ -65,6 +65,20 @@ object OptimizerTest extends JasmineTest {
       expect(elements).toEqual(js.Array(10, 9, 8, 7, 6, 5, 4, 3, 2))
     }
 
+    it("must not break classOf[T] == classOf[U] - #1658") {
+      expect(classOf[String] == classOf[String]).toBeTruthy
+      expect(classOf[Int] == classOf[Int]).toBeTruthy
+      expect(classOf[Array[Int]] == classOf[Array[Int]]).toBeTruthy
+      expect(classOf[Array[String]] == classOf[Array[String]]).toBeTruthy
+
+      expect(classOf[String] == classOf[Int]).toBeFalsy
+      expect(classOf[Seq[_]] == classOf[List[_]]).toBeFalsy
```

Comment:
At the `j.l.Class` level, it's pretty irrelevant how they end up being represented in JS.

For tests relevant to JS primitive representation, see https://github.com/scala-js/scala-js/blob/master/test-suite/src/test/scala/org/scalajs/testsuite/compiler/InstanceTestsHijackedBoxedClassesTest.scala

But I can add `classOf[Int] == classOf[Long]` if you want ;)